### PR TITLE
Shutdown duckdb on disconnect from R by default

### PR DESCRIPTION
**Pull request by ablack3**
Issue discussed in https://github.com/duckdb/duckdb/issues/3451.

### CODE REVIEW

**Comment by hannes** [comment]
I think this makes sense because most people only ever are going to have a single connection.

**Comment by hannes** [comment]
Can you have a look at the failing tests please?

**Comment by ablack3** [comment]
I found the problem.

`dbDisconnect(conn, shutdown = TRUE)` will call `duckdb_shutdown(conn@driver)` which will call `dbIsValid(drv)` which gets dispatched to `dbIsValid__duckdb_driver` which calls `dbDisconnect` using the default shutdown argument. 

By changing the default shutdown argument to TRUE I've created a recursive call causing the C stack overload error.

The fix I applied was to change the `dbDisconnect(conn)` call in dbIsValid__duckdb_driver to `dbDisconnect(conn, shutdown = FALSE)`. This avoids the recursion and keeps the behavior of dbIsValid__duckdb_driver exactly as it was before. I'm not sure this is the best solution though. `dbDisconnect` is still calling `dbDisconnect`.

**Comment by ablack3** [comment]
I made a few more changes to get tests passing.

File: tools/rpkg/tests/testthat/test-DBItest.R [hunk 1]
```diff
@@ -43,4 +43,6 @@ DBItest::test_all(c(
   "bind_factor",
 
   NULL
-))
+  ),
+  ctx <- DBItest::make_context(new("DBIConnector", .drv = duckdb(), .conn_args = list(dbdir = ":memory:")))
```
Comment (ablack3) [inline tools/rpkg/tests/testthat/test-DBItest.R:47]: I had to be explicit about creating the context for DBITest. Not sure why but the default context was causing an error.

File: tools/rpkg/R/dbDisconnect__duckdb_connection.R [hunk 2]
```diff
@@ -6,7 +6,7 @@
 #' @param shutdown Set to `TRUE` to shut down the DuckDB database instance that this connection refers to.
 #' @rdname duckdb
 #' @usage NULL
-dbDisconnect__duckdb_connection <- function(conn, ..., shutdown = FALSE) {
+dbDisconnect__duckdb_connection <- function(conn, ..., shutdown = TRUE) {
```
Comment (ablack3) [inline tools/rpkg/R/dbDisconnect__duckdb_connection.R:9]: The ` duckdb_shutdown(conn@driver)` call in this function will call `dbDisconnect__duckdb_connection` creating the potential for infinite recursion.

**Comment by ablack3** [comment]
The current error seems to have nothing to do with my changes. 

image

File: tools/rpkg/R/dbDisconnect__duckdb_connection.R (hunk 2 above)
Comment (ablack3) [reply tools/rpkg/R/dbDisconnect__duckdb_connection.R:9]: I think I noticed some weird behavior with this branch. I'm thinking maybe the recursive call should be avoided.

**Comment by eitsupi** [comment]
Any update?

**Comment by Mytherin** [comment]
@hannes perhaps have another look at this one?

**Comment by hannes** [comment]
@krlmlr any opinion about this?

**Comment by ablack3** [comment]
I'm open to putting more time on this if you have specific suggestions to improve the PR. Thanks for considering it!

**Comment by krlmlr** [comment]
Thanks. The implementation seems to touch many things, I haven't reviewed in detail.

The original issue describes a file access error, can that be resolved with a clean `dbDisconnect()` ? The following is with a recent version of duckdb, and seems to work without shutting down duckdb.

``` r
con <- DBI::dbConnect(duckdb::duckdb(), "test.duckdb")

DBI::dbCreateTable(con, "test", data.frame(a = 1))
DBI::dbListTables(con)
#> [1] "test"

DBI::dbDisconnect(con, shutdown = FALSE)

con <- DBI::dbConnect(duckdb::duckdb(), "test.duckdb")
DBI::dbCreateTable(con, "test2", data.frame(a = 1))
DBI::dbListTables(con)
#> [1] "test"  "test2"
```

<sup>Created on 2023-02-23 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

What are the costs for spinning up and shutting down the duckdb engine? What happens if a session creates two in-memory connections, or two connections to different files?

**Comment by krlmlr** [comment]
I don't follow.

**Comment by ablack3** [comment]
@krlmlr One thing I'm running into a lot is that I want to disconnect a DBI connection and it could be duckdb or it could be another connection type. 

So I need logic like

```
if (class(conn) == "duckdb_connection") {
  DBI::dbDisconnect(conn, shutdown = TRUE)
} else {
  DBI::dbDisconnect(conn)
}
```

Maybe I'm misunderstanding something.

**Comment by krlmlr** [comment]
Are you trying to get rid of the warning:

```
Warning: Database is garbage-collected, use dbDisconnect(con, shutdown=TRUE) or duckdb::duckdb_shutdown(drv) to avoid this.
```

?

The way I see it, there are two main use cases:

## Anonymous `duckdb()` driver

```r
library(duckdb)

options(warn = 1)

for (i in 1:3) {
  con <- dbConnect(duckdb())
  dbDisconnect(con)
  con <- NULL
  gc()
}
```

This shows three warnings.

## Allocate and keep driver

```r
library(duckdb)

drv <- duckdb()

for (i in 1:3) {
  con <- dbConnect(drv)
  dbDisconnect(con)
  con <- NULL
  gc()
}

duckdb_shutdown(drv)
```

This doesn't show warnings thanks to the `duckdb_shutdown()`

## Proposed solution

I believe we can get rid of the warning in the first example, without changing the default of the `shutdown` argument. In R there's a way to check if an object has been aliased (duplicated). In the first example, it won't be duplicated, and hence safe to shutdown, even if the user had specified `shutdown = FALSE` (the default). This would get rid of the warning.

Do I understand you correctly? What do you think?

**Comment by rsund** [comment]
On **Windows** the following error is seen (with latest master installed from r-universe) when trying to reopen connection to the same **file-based** database:
```r
> con <- DBI::dbConnect(duckdb::duckdb(), dbdir="test.db")
> DBI::dbDisconnect(con)
> con <- DBI::dbConnect(duckdb::duckdb(), dbdir="test.db")
Error: rapi_startup: Failed to open database: IO Error: Cannot open file "test.db": The process cannot access the file because it is being used by another process.
```

This can be avoided by using `DBI::dbDisconnect(con, shutdown=TRUE)` [that actually calls `duckdb::duckdb_shutdown()`](https://github.com/duckdb/duckdb/blob/a12329e657df33c4e47e62459e42760f1fe49610/tools/rpkg/R/dbDisconnect__duckdb_connection.R#L16).

The file opening error can also be avoided by using `con <- NULL; gc()` after `DBI::dbDisconnect(con)` which also [deletes the driver in the finaliser](https://github.com/duckdb/duckdb/blob/a12329e657df33c4e47e62459e42760f1fe49610/tools/rpkg/src/database.cpp#L14) and prints the warning `Database is garbage-collected, use dbDisconnect(con, shutdown=TRUE) or duckdb::duckdb_shutdown(drv) to avoid this.`

In any case, the driver must be shut down in a way or another to avoid **file** access error on **Windows**.

So I think that changing the default to `shutdown=TRUE` is needed and what the users expect. The problem seems to be **Windows** related, but as `shutdown=TRUE` is not causing problems on Mac or Linux, it should be the default as the `shutdown=FALSE` is not working on Windows.

However, giving a warning in the case that there are aliased objects detected while shutting the driver down is certainly a good idea as then the user knows to use `shutdown=FALSE` if that is really sometimes needed. I have difficulties to see that it would be needed at the R level, but I can certainly be wrong and it would be nice to see real use cases that really require `shutdown=FALSE`.

**Comment by krlmlr** [comment]
I was missing the "Windows only" part. I agree that the current behavior isn't nice.

I'll do some testing and report here.

**Comment by krlmlr** [comment]
This is an example that shows how we could detect if the duckdb instance behind the connection is safe to shut down. The "connection" in this example is just a flag that indicates if the "duckdb instance" has been bound to another symbol when this was called.

@hannes: Should we follow this route?

``` r
cpp11::cpp_source(
  code = '#include "cpp11/list.hpp"

  [[cpp11::register]]
  SEXP connect(SEXP x) {
    auto shared = MAYBE_SHARED(x);
    SEXP out = Rf_allocVector(LGLSXP, 1);
    LOGICAL(out)[0] = shared;
    return out;
  }
  '
)

duckdb <- function() {
  list()
}

disconnect <- function(con) {
  if (con) {
    message("May be shared, can't shut down")
  } else {
    message("Not shared, safe to disconnect")
  }
}

con1 <- connect(duckdb())
disconnect(con1)
#> Not shared, safe to disconnect

drv1 <- duckdb()
con2 <- connect(drv1)
disconnect(con2)
#> May be shared, can't shut down

con2 <- connect(drv2 <- duckdb())
disconnect(con2)
#> May be shared, can't shut down
```

<sup>Created on 2023-03-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by Mytherin** [comment]
Closing as this has been superseded by #8273

**Pull request closed without merging**